content/docs/render-props.md

```diff
@@ -0,0 +1,287 @@
+
+More concretely, **a render prop is a function prop that a component uses to know what to render.**
+
+This technique makes the behavior that we need to share extremely portable. To get that behavior, render a `<Mouse>` with a `render` prop that tells it what to render with the current (x, y) of the cursor.
+
+One interesting thing to note about render props is that you can implement most [higher-order components](/docs/higher-order-components.html) (HOC) using a regular component with a render prop. For example, if you would prefer to have a `withMouse` HOC instead of a `<Mouse>` component, you could easily create one using a regular `<Mouse>` with a render prop:
+
+```js
+// If you really want a HOC for some reason, you can easily
+// create one using a regular component with a render prop!
+function withMouse(Component) {
+  return class extends React.Component {
+    render() {
+      return (
+        <Mouse render={mouse => (
+          <Component {...this.props} mouse={mouse} />
+        )}/>
+      );
+    }
+  }
+}
+```
+
+So using a render prop makes it possible to use either pattern.
+
+## Caveats
+
+### Be careful when using Render Props with React.PureComponent
+
+Using a render prop can negate the advantage that comes from using [`React.PureComponent`](/docs/react-api.html#reactpurecomponent) if you create the function inside a `render` method. This is because the shallow prop comparison will always return `false` for new props, and each `render` in this case will generate a new value for the render prop.
+
+For example, continuing with our `<Mouse>` component from above, if `Mouse` were to extend `React.PureComponent` instead of `React.Component`, our example would look like this:
+
+```js
+class Mouse extends React.PureComponent {
+  // Same implementation as above...
+}
+
+class MouseTracker extends React.Component {
+  render() {
+    return (
+      <div>
+        <h1>Move the mouse around!</h1>
+
+        {/*
+          This is bad! The value of the `render` prop will
+          be different on each render.
+        */}
+        <Mouse render={mouse => (
+          <Cat mouse={mouse} />
+        )}/>
+      </div>
+    );
+  }
+}
+```
+
+In this example, each time `<MouseTracker>` renders, it generates a new function as the value of the `<Mouse render>` prop, thus negating the effect of `<Mouse>` extending `React.PureComponent` in the first place!
+
+To get around this problem, you can sometimes define the prop as an instance method, like so:
+
+```js
+class MouseTracker extends React.Component {
+  constructor(props) {
+    super(props);
+
+    // This binding ensures that `this.renderTheCat` always refers
+    // to the *same* function when we use it in render.
+    this.renderTheCat = this.renderTheCat.bind(this);
```

Comment:
I think we often see the usage of arrow functions instead of binding in the constructor. 
Is there any particular reason not to use an arrow function?

Reply:
Yes. The whole point of this example is that we create the function only once instead of dynamically creating a function on every render. Unless you're suggesting I use an arrow function here in the constructor?

Reply:
No, I meant Instead of this:
```
class MouseTracker extends React.Component {
  constructor(props) {
    super(props);

    // This binding ensures that `this.renderTheCat` always refers
    // to the *same* function when we use it in render.
    this.renderTheCat = this.renderTheCat.bind(this);
  }

  renderTheCat(mouse) {
    return <Cat mouse={mouse} />
  }

  render() {
    return (
      <div>
        <h1>Move the mouse around!</h1>
        <Mouse render={this.renderTheCat} />
      </div>
    );
  }
}
```

*Maybe* this makes more sense/more consistent?
```
class MouseTracker extends React.Component {
   // `this.renderTheCat` always refers to the *same* function when we use it in render.
  renderTheCat = (mouse) => {
    return <Cat mouse={mouse} />
  }

  render() {
    return (
      <div>
        <h1>Move the mouse around!</h1>
        <Mouse render={this.renderTheCat} />
      </div>
    );
  }
}
```

Reply:
The docs intentionally try to stick to standard syntax instead of proposal-stage. Class properties are getting there but aren't standard just yet 😄 

https://github.com/reactjs/reactjs.org/blob/master/CONTRIBUTING.md#dont-use-features-that-arent-standardized-yet

Reply:
@alexkrolick my bad. I wasn't aware it isn't standard yet (with babel, the difference between standard and non standard is blurry). Thanks for the clarification.

Reply:
FWIW @ron23 I always write my code in the way you're suggesting here and never actually write out my `constructor`s, so I feel you. But ya, the React docs tend to favor `.bind` at present.

Reply:
> // This binding ensures that `this.renderTheCat` always refers
   // to the *same* function when we use it in render.

Is there any case that `this.renderTheCat` refers a *different* function in render without this?
I know that `this` in the function may not be a MouseTracker's instance, but I guess the reference to the function is always the same.